File: laby/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

import java.util.Objects;

public class Vector2d {
    private final int x;
    private final  int y;
    public Vector2d(int x, int y) {
        this.x = x;
        this.y = y;
    }
    public int getX() {
        return this.x;
    }
    public int getY() {
        return this.y;
    }
    public String toString() {
        return "(" + this.x + ", " + this.y + ")";
    }
    public boolean precedes(Vector2d vector) {
        return this.x <= vector.getX() && this.y <= vector.getY();
    }
    public boolean followes(Vector2d vector) {
        return this.x >= vector.getX() && this.y >= vector.getY();
    }
    public Vector2d add(Vector2d vector) {
        return new Vector2d(x + vector.getX(), y + vector.getY());
    }
    public Vector2d subtract(Vector2d vector) {
        return new Vector2d(x - vector.getX(), y - vector.getY());
    }
    public Vector2d upperRight(Vector2d vector) {
        return new Vector2d( Math.max(x, vector.getX()), Math.max(y, vector.getY()));
    }
    public Vector2d lowerLeft(Vector2d vector) {
        return new Vector2d( Math.min(x, vector.getX()), Math.min(y, vector.getY()));
    }
    public Vector2d opposite() {
        return new Vector2d(-this.x, -this.y);
    }
    public boolean equals(Vector2d vector) {
        if (this==vector) return true;
        if (vector == null) return false;
        return this.x == vector.getX() && this.y == vector.getY();
    }
    public int hashCode() {
        return Objects.hash(this.x,this.y);
    }




}

File: labK/src/main/kotlin/agh/isc/oop/model/Vector2d.kt
package agh.ics.oop.model

import main.agh.ics.oop.model.MapDirection

data class Vector2d(val x:Int,  val y: Int)
{
    operator fun compareTo(other : Vector2d): Int {
       return when{
           this.x >= other.x && this.y >= other.y -> 1
              this.x <= other.x && this.y <= other.y -> -1
              else -> 0
       }

    }
    operator  fun unaryMinus(): Vector2d = Vector2d(-this.x, -this.y)


    operator fun plus(other: Vector2d): Vector2d = Vector2d(this.x + other.x, this.y + other.y)


    operator fun minus(other: Vector2d): Vector2d = Vector2d(this.x - other.x, this.y - other.y)


    public fun upperRight(other: Vector2d): Vector2d =Vector2d(Math.max(this.x, other.x), Math.max(this.y, other.y))


    public fun lowerLeft(other: Vector2d): Vector2d =Vector2d(Math.min(this.x, other.x), Math.min(this.y, other.y))







}

fun MapDirection.toUnitVector(): Vector2d {
    return when (this) {
        MapDirection.NORTH -> Vector2d(0, 1)
        MapDirection.EAST -> Vector2d(1, 0)
        MapDirection.SOUTH -> Vector2d(0, -1)
        MapDirection.WEST -> Vector2d(-1, 0)
    }
}
